src/ServiceControl.Transports.SQLServer/NativeQueueLengthProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿namespace ServiceControl.Transports.SQLServer
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Data.SqlClient;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.QueueLength;
+
+    public class NativeQueueLengthProvider : INativeQueueLengthProvider
+    {
+        string connectionString;
+
+        ConcurrentDictionary<string, EndpointInstanceId> queueNames = new ConcurrentDictionary<string, EndpointInstanceId>();
+        ConcurrentDictionary<string, int> queueSizes = new ConcurrentDictionary<string, int>();
+        Task pooler;
+        CancellationTokenSource stop;
+
+        public Dictionary<EndpointInstanceId, int> SnapshotValues()
+        {
+            var result = queueNames
+                .ToLookup(kv => kv.Value, kv => kv.Key)
+                .ToDictionary(g => g.Key, g => g.ToArray().Sum(qn =>
+                                                            {
+                                                                queueSizes.TryGetValue(qn, out var length);
+                                                                return length;
+                                                            })
+                );
+
+            return result;
+        }
```

Comment:
I think we'll still need to break this down on EndpointInstanceId _and_ physical queue name. If we don't then we'll double count shared queues in competing consumer mode.